### PR TITLE
TypeScript: Strict null checks

### PR DESCRIPTION
**Pull request by jpbernius**
This Pull Request changes the TypeScript Compiler to strict null checks.
Also, it fixes the compile errors originating from this setting change.

### Checklist
- [x] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [X] ~I documented my source code using the JavaDoc / JSDoc style.~ (No Changes beside Type changes.)
- [X] ~I added integration test cases for the server (Spring) related to the features.~ (No Server Side Changes)
- [X] ~I added integration test cases for the client (Jest) related to the features.~ (No change in clientside functionality)
- [X] ~I added screenshots/screencast of my UI changes~ (No UI changes)
- [X] ~I translated all the newly inserted strings~ (No string changes)

### CODE REVIEW

File: src/main/webapp/app/core/websocket/websocket.service.ts [hunk 1]
```diff
@@ -137,6 +137,7 @@ export class JhiWebsocketService implements OnDestroy {
         if (channel != null && (!Object.keys(this.myListeners).length || !this.myListeners.hasOwnProperty(channel))) {
             this.myListeners[channel] = this.createListener(channel);
         }
+        // TODO: Fix access of myListeners[undefined]
```
Comment (jpbernius) [inline src/main/webapp/app/core/websocket/websocket.service.ts]: Can someone help me with this fix access of `myListeners[undefined]`?

Comment (jpbernius) [inline src/main/webapp/app/core/websocket/websocket.service.ts]: Can someone help me with this fix access of `myListeners[undefined]`?

File: src/main/webapp/app/core/websocket/websocket.service.ts [hunk 2]
```diff
@@ -166,6 +167,7 @@ export class JhiWebsocketService implements OnDestroy {
             if (channel != null && (!Object.keys(this.myListeners).length || !this.myListeners.hasOwnProperty(channel))) {
                 this.myListeners[channel] = this.createListener(channel);
             }
+            // TODO: Fix access of myListeners[undefined]
```
Comment (jpbernius) [inline src/main/webapp/app/core/websocket/websocket.service.ts]: Can someone help me with this fix access of `myListeners[undefined]`?

File: src/main/webapp/app/core/websocket/websocket.service.ts [hunk 3]
```diff
@@ -180,6 +182,7 @@ export class JhiWebsocketService implements OnDestroy {
     }
 
     unsubscribe(channel?: string) {
+        // TODO: Fix access of myListeners[undefined]
```
Comment (jpbernius) [inline src/main/webapp/app/core/websocket/websocket.service.ts]: Can someone help me with this fix access of `myListeners[undefined]`?

File: src/main/webapp/app/entities/lecture/lecture-attachments.component.ts [hunk 4]
```diff
@@ -59,26 +59,27 @@ export class LectureAttachmentsComponent implements OnInit {
 
     saveAttachment() {
         if (!this.attachmentToBeCreated) {
+            // TODO: Fix this Null Pointer, maybe call addAttachment()?
```
Comment (jpbernius) [inline src/main/webapp/app/entities/lecture/lecture-attachments.component.ts]: @maximilianmeier Can you look at this Null Pointer?

File: src/main/webapp/app/entities/result/result-history.component.ts [hunk 5]
```diff
@@ -36,7 +36,8 @@ export class ResultHistoryComponent {
             return 0;
         }
         if (result.resultString && result.resultString.indexOf('failed') !== -1) {
-            return null;
+            // TODO: Check if returning 0 is ok
```
Comment (jpbernius) [inline src/main/webapp/app/entities/result/result-history.component.ts]: @maximilianmeier Can you check if returning 0 is ok?

File: src/main/webapp/app/overview/course-statistics/course-statistics.component.ts [hunk 6]
```diff
@@ -267,7 +267,8 @@ export class CourseStatisticsComponent implements OnInit, OnDestroy {
             return 0;
         }
         if (result.resultString && result.resultString.indexOf('failed') !== -1) {
-            return null;
+            // TODO: Is this right?
```
Comment (jpbernius) [inline src/main/webapp/app/overview/course-statistics/course-statistics.component.ts]: @maximilianmeier Can you check if returning 0 is ok?

File: src/main/webapp/app/tutor-exercise-dashboard/tutor-exercise-dashboard.component.ts [hunk 7]
```diff
@@ -293,6 +294,7 @@ export class TutorExerciseDashboardComponent implements OnInit {
             route = `/modeling-exercise/${this.exercise.id}/submissions/${submission}/assessment`;
             queryParams.showBackButton = true;
         }
+        // TODO: Make sure route is set
```
Comment (jpbernius) [inline src/main/webapp/app/tutor-exercise-dashboard/tutor-exercise-dashboard.component.ts]: @SiggZ Can you make sure `route` is set?

**Comment by jpbernius** [comment]
@thilo-behnke Can you check out the Errors in `build-log.model.ts` and `code-editor-ace.component.ts`?

**Comment by thilo-behnke** [comment]
@jpbernius I have fixed the typing issues in the code-editor components.

File: src/main/webapp/app/entities/lecture/lecture-attachments.component.ts (hunk 4 above)
Comment (maximilianmeier) [reply src/main/webapp/app/entities/lecture/lecture-attachments.component.ts]: done

**Comment by maximilianmeier** [comment]
we have a import issue right now `import { ProfileService, ProfileInfo } from 'app/layouts';`
 this causes a problem like `Uncaught Error: Can't resolve all parameters for NavbarComponent: ([object Object], [object Object], [object Object], [object Object], [object Object], ?, [object Object]).` in several components

File: src/main/webapp/app/core/websocket/websocket.service.ts (hunk 3 above)
Comment (maximilianmeier) [reply src/main/webapp/app/core/websocket/websocket.service.ts]: i removed the optional. this caused only one problem in the app. i fixed that issue

File: src/main/webapp/app/core/websocket/websocket.service.ts (hunk 2 above)
Comment (maximilianmeier) [reply src/main/webapp/app/core/websocket/websocket.service.ts]: i removed the optional. this caused only one problem in the app. i fixed that issue

File: src/main/webapp/app/core/websocket/websocket.service.ts (hunk 1 above)
Comment (maximilianmeier) [reply src/main/webapp/app/core/websocket/websocket.service.ts]: i removed the optional. this caused only one problem in the app. i fixed that issue

File: src/main/webapp/app/entities/result/result-history.component.ts (hunk 5 above)
Comment (maximilianmeier) [reply src/main/webapp/app/entities/result/result-history.component.ts]: we need the null. i reverted it

File: src/main/webapp/app/overview/course-statistics/course-statistics.component.ts (hunk 6 above)
Comment (maximilianmeier) [reply src/main/webapp/app/overview/course-statistics/course-statistics.component.ts]: we need the null. i reverted it

File: src/main/webapp/app/entities/result/result-history.component.ts (hunk 5 above)
Comment (jpbernius) [reply src/main/webapp/app/entities/result/result-history.component.ts]: Did you forget to push that change? 😀

**Comment by jpbernius** [comment]
I reverted the change of the import that causes problems for you @maximilianmeier in 
61e2f9c.

File: src/main/webapp/app/tutor-exercise-dashboard/tutor-exercise-dashboard.component.ts (hunk 7 above)
Comment (jpbernius) [reply src/main/webapp/app/tutor-exercise-dashboard/tutor-exercise-dashboard.component.ts]: I added a fix in ce859b5. Can you verify that it is ok?

Comment (SiggZ) [reply src/main/webapp/app/tutor-exercise-dashboard/tutor-exercise-dashboard.component.ts]: The fix looks good to me.

File: src/main/webapp/app/entities/result/result-history.component.ts (hunk 5 above)
Comment (maximilianmeier) [reply src/main/webapp/app/entities/result/result-history.component.ts]: yeah. i pushed it

**Comment by krusche** [comment]
We will integrate this into the next major release 3.2.0 as we need more testing.
Before that, we will have additional bugfix releases to make sure the current functionality works as stable as possible

**Comment by krusche** [comment]
Sorry, I did not have time yet to review this. I suggest we add this to the next big release 3.2.0.
Would be great if you could fix the merge conflicts and we review this together after WWDC, when we are both back in Munich

**Comment by krusche** [comment]
@jpbernius What is the status of this PR? Do we have any progress? Would be great if we can either finish and integrate it in the next days or close it.

**Pull request merged**